File: web-backend/src/services/serviceAccountService.py
from google.cloud import iam, artifactregistry_v1
from google.cloud.iam_admin_v1.services.iam import IAMClient
from google.cloud.iam_admin_v1.types import CreateServiceAccountRequest, ServiceAccount
from google.cloud.artifactregistry_v1 import ArtifactRegistryClient
from google.cloud.artifactregistry_v1.types import Repository

def create_service_account(iam_client, project_id, service_account_id):
    """Create a new service account."""
    service_account = iam_client.create_service_account(
        request=CreateServiceAccountRequest(
            name=f"projects/{project_id}",
            account_id=service_account_id,
            service_account=ServiceAccount(display_name=github_username)
        )
    )
    print(f"Created service account: {service_account.email}")
    return service_account.email

def grant_artifact_registry_admin_role(iam_client, project_id, service_account_email):
    """Grant Artifact Registry Admin role to the service account."""
    policy = iam_client.get_iam_policy(resource=f"projects/{project_id}")
    binding = next((b for b in policy.bindings if b.role == "roles/artifactregistry.admin"), None)
    if binding:
        if service_account_email not in binding.members:
            binding.members.append(f"serviceAccount:{service_account_email}")
    else:
        policy.bindings.append({"role": "roles/artifactregistry.admin", "members": [f"serviceAccount:{service_account_email}"]})
    iam_client.set_iam_policy(resource=f"projects/{project_id}", policy=policy)
    print(f"Granted Artifact Registry Admin role to {service_account_email}")

def create_artifact_registry_repo(artifact_client, project_id, location, repository_id):
    """Create a new Artifact Registry repository."""
    parent = f"projects/{project_id}/locations/{location}"
    repository = Repository(
        name=f"{parent}/repositories/{repository_id}",
        format_=Repository.Format.DOCKER,
        description="GitHub Actions Docker Repository"
    )
    response = artifact_client.create_repository(parent=parent, repository_id=repository_id, repository=repository)
    print(f"Created Artifact Registry repository: {response.name}")

def main(github_username):
    project_id = "your-project-id"
    location = "us-east4"  # e.g., "us-central1"
    repository_id = "your-repo-id"

    # Initialize clients
    iam_client = IAMClient()
    artifact_client = ArtifactRegistryClient()

    # Step 1: Create a service account
    service_account_email = create_service_account(iam_client, project_id, github_username)

    # Step 2: Grant Artifact Registry Admin role to the service account
    grant_artifact_registry_admin_role(iam_client, project_id, service_account_email)

    # Step 3: Create an Artifact Registry repository
    create_artifact_registry_repo(artifact_client, project_id, location, repository_id)

    # Steps related to Workload Identity Federation and connecting to GitHub are not
    # directly executable in this script and would involve manual steps or using gcloud commands.
    # This would include creating a workload identity pool, a provider, and binding the GitHub repository
    # to the service account with specific roles.

if __name__ == "__main__":
    github_username = "bobofishbo"  # Replace with your GitHub username
    main(github_username)

File: web-backend/src/services/serviceAccountService.ts
import { exec } from 'child_process'
import { writeFile } from 'fs/promises'
import { tmpdir } from 'os'
import { join } from 'path'
import { promisify } from 'util'

const execAsync = promisify(exec)

async function logCurrentServiceAccount() {
  try {
    const { stdout } = await execAsync('gcloud config get-value account')
    const lines = stdout.split('\n').filter(line => line.trim() !== '')
    const accountEmail = lines[lines.length - 1].trim()
    console.log(`Current service account: ${accountEmail}`)
  } catch (error) {
    console.error('Error fetching current service account:', error)
  }
}

async function createServiceAccount(
  projectId: string,
  serviceAccountId: string,
  displayName: string,
) {
  const command = `gcloud iam service-accounts create ${serviceAccountId} --project=${projectId} --display-name="${displayName}"`
  await execAsync(command)
  console.log(`Created service account: ${serviceAccountId}`)
}

async function bindGitHubRepoToServiceAccount(
  projectId: string,
  serviceAccountEmail: string,
  githubRepoRef: string,
  workloadIdentityPoolId: string,
) {
  const command = `gcloud iam service-accounts add-iam-policy-binding "${serviceAccountEmail}" --project="${projectId}" --role="roles/iam.workloadIdentityUser" --member="principalSet://iam.googleapis.com/${workloadIdentityPoolId}/attribute.repository/${githubRepoRef}"`
  await execAsync(command)
  console.log(
    `Bound GitHub repository ${githubRepoRef} to service account ${serviceAccountEmail}`,
  )
}

async function createArtifactRegistryRepo(
  projectId: string,
  location: string,
  repositoryId: string,
) {
  const createCommand = `gcloud artifacts repositories create ${repositoryId} --project=${projectId} --location=${location} --repository-format=docker`
  await execAsync(createCommand)
  console.log(`Created Artifact Registry repository: ${repositoryId}`)

  const policyContent = [
    {
      name: 'delete-untagged-artifacts',
      action: {
        type: 'Delete',
      },
      condition: {
        tagState: 'untagged',
      },
    },
  ]

  const tempDir = tmpdir()
  const policyFile = join(tempDir, 'cleanup_policy.json')

  await writeFile(policyFile, JSON.stringify(policyContent, null, 2))

  const setPolicyCommand = `gcloud artifacts repositories set-cleanup-policies ${repositoryId} --project=${projectId} --location=${location} --policy=${policyFile}`
  await execAsync(setPolicyCommand)
  console.log(`Set cleanup policy for repository: ${repositoryId}`)
}

async function grantArtifactRegistryWriterRole(
  projectId: string,
  location: string,
  repositoryId: string,
  serviceAccountEmail: string,
) {
  const command = `gcloud artifacts repositories add-iam-policy-binding ${repositoryId} --project=${projectId} --location=${location} --member="serviceAccount:${serviceAccountEmail}" --role="roles/artifactregistry.writer"`
  await execAsync(command)
  console.log(
    `Granted Artifact Registry Writer role to ${serviceAccountEmail} for repository ${repositoryId}`,
  )
}

async function createServiceAccountAndResources(githubUsername: string) {
  const username = githubUsername.toLowerCase().replace(/[^a-z0-9-]/g, '')
  const projectId = 'pokerai-417521'
  const serviceAccountId = username
  const displayName = username
  const serviceAccountEmail = `${username}@${projectId}.iam.gserviceaccount.com`
  const githubRepoRef = `${githubUsername}/poker-engine-2024`
  const workloadIdentityPoolId =
    'projects/979321260256/locations/global/workloadIdentityPools/github'
  const location = 'us-east4'
  const repositoryId = username

  try {
    await logCurrentServiceAccount()
    // Step 1: Create a service account
    await createServiceAccount(projectId, serviceAccountId, displayName)

    // Step 2: Bind the GitHub repository to the service account
    await bindGitHubRepoToServiceAccount(
      projectId,
      serviceAccountEmail,
      githubRepoRef,
      workloadIdentityPoolId,
    )

    // Step 3: Create an Artifact Registry repository
    await createArtifactRegistryRepo(projectId, location, repositoryId)

    // Step 4: Grant the Artifact Registry Writer role to the service account for the specific repository
    await grantArtifactRegistryWriterRole(
      projectId,
      location,
      repositoryId,
      serviceAccountEmail,
    )

    console.log('Service account and resources setup completed successfully.')
  } catch (error) {
    console.error('Error setting up the service account and resources:', error)
  }
}

export default createServiceAccountAndResources
